### PR TITLE
Make input partitions methods based on asset key not input name

### PR DESCRIPTION
**Pull request by jamiedemaria**
## Summary & Motivation
One of the changes I'd like to make to the partition methods on the context is to allow them to accept `CoercibleToAssetKey` rather than input name. This will mean users wont need to map back and forth between asset key and input name and will remove the I/O manager-centric term "input". 

To achieve this, the top level partition functions in the StepExecutionContext need to accept asset keys rather then input name. Right now the pattern is like this 

```python
class OpExecutionContext:

   def asset_partition_key_for_input(input_name):
        self.step_execution_context.asset_partition_key_for_input(input_name)

class StepExecutionContext:
    def asset_partition_key_for_input(input_name):
         asset_key = self.get_asset_key_for_input(input_name)
         ...
```

And I'd like it to be like this:
```python
class OpExecutionContext:

   def asset_partition_key_for_input(input_name):
       asset_key = self.step_execution_context.get_asset_key_for_input(input_name)
        self.step_execution_context.asset_partition_key_for_asset(asset_key)

class StepExecutionContext:
    def asset_partition_key_for_asset(asset_key):
         ...
```

So that we can support this in `AssetExecutionContext`

```python
class AssetExecutionContext:

   def partition_key_for_upstream_asset(asset_key: CoercibleToAssetKey):
        self.step_execution_context.asset_partition_key_for_asset(AssetKey.from_coercible(asset_key))
```

### CODE REVIEW

File: python_modules/dagster/dagster/_core/execution/context/system.py [hunk 1]
```diff
@@ -1109,22 +1112,16 @@ def get_output_asset_keys(self) -> AbstractSet[AssetKey]:
             output_keys.add(asset_info.key)
         return output_keys
 
-    def has_asset_partitions_for_input(self, input_name: str) -> bool:
```
Comment (jamiedemaria) [inline python_modules/dagster/dagster/_core/execution/context/system.py:1112]: What's the deprecation policy for these methods on StepExecutionContext? Are they considered public? If so i'll retain versions of these methods that just call their asset_key based counterparts after converting the input name to an asset_key

Comment (sryza) [reply python_modules/dagster/dagster/_core/execution/context/system.py:1112]: Methods are only public if:
- They're on a publicly-exported class
- They have the `@public` decorator

Comment (jamiedemaria) [reply python_modules/dagster/dagster/_core/execution/context/system.py:1112]: ok - these are public then. I'll add the method names back

Comment (sryza) [reply python_modules/dagster/dagster/_core/execution/context/system.py:1112]: This one at least doesn't have the `@public` decorator, right?

Comment (jamiedemaria) [reply python_modules/dagster/dagster/_core/execution/context/system.py:1112]: yeah, but StepExecutionContext is publicly exported class.

Comment (jamiedemaria) [reply python_modules/dagster/dagster/_core/execution/context/system.py:1112]: wait... was the list of conditions an AND not an OR?

Comment (sryza) [reply python_modules/dagster/dagster/_core/execution/context/system.py:1112]: ah oops yes, and AND

Comment (jamiedemaria) [reply python_modules/dagster/dagster/_core/execution/context/system.py:1112]: ok i can revert that commit lol

**Comment by jamiedemaria** [comment]
converting back to draft to get this out of review queues until context work is re-prioritized

**Pull request closed without merging**